### PR TITLE
User profile notification

### PR DESCRIPTION
**Pull request by tpan501**
the later test files will be uploaded to this branch.
you can merge it first to get nofitication function (most of it)

### CODE REVIEW

File: backend/src/routes/api/group.js [hunk 1]
```diff
@@ -368,11 +480,45 @@ router.patch("/dismiss/:groupId", async (req, res) => {
...
+      const notificationType = "group_dismissed";
+      const notificationPromises = userIds.map(async (id) => {
+        const newNotification = new Notification({
+          notificationContent,
+          isRead: false,
+          notificationTime,
+          notificationType,
+          senderId: userId,
+          receiverId: id,
+        });
```
Comment (ryanwooa) [inline backend/src/routes/api/group.js:549]: Add group_id of notification in your branch. Edit the notification schema
If not, you will need to edit all the new notification statement after Kelvin add in his branch.

File: backend/src/routes/api/group.js [hunk 2]
```diff
@@ -167,6 +180,24 @@ router.patch("/update/:id", getGroup, async (req, res) => {
...
+        notificationTime,
+        notificationType,
+        senderId: req.user._id,
+        receiverId: id,
+      });
+      return newNotification.save();
+    });
+
+    await Promise.all(notificationPromises);
+
```
Comment (ryanwooa) [inline backend/src/routes/api/group.js:177]: This a update route, you sure this only for close?

File: backend/src/routes/api/group.js [hunk 3]
```diff
@@ -230,8 +283,10 @@ router.post("/join/:id", getGroup, async (req, res) => {
 
 // quit group by id
 router.post("/quit/:groupId", async (req, res) => {
+  console.log("Quit group route");
   const { groupId } = req.params;
   const userId = req.user._id; // User ID from authentication/session
+  const user = await User.findById(userId);
```
Comment (ryanwooa) [inline backend/src/routes/api/group.js:268]: req.user already is the user date, why need to finById one more time?

File: backend/src/routes/api/group.js [hunk 4]
```diff
@@ -254,38 +325,74 @@ router.post("/quit/:groupId", async (req, res) => {
...
       applicationStatus: "pending",
-      applicationDate: new Date(),
+      applicationDate: new Date()
     });
     await newApplication.save();
 
-    res.group.application.push(newApplication._id);
+    //create a new notification
+    const newNotification = new Notification({
+      notificationContent: `New application from ${user.name} to join group ${res.group.groupName}`,
```
Comment (ryanwooa) [inline backend/src/routes/api/group.js]: How about change to :   Group name + apply message

**Review by ryanwooa — APPROVED** [review]

**Review by zli213 — APPROVED** [review]

**Review by xinranwangxwan559 — APPROVED** [review]

**Pull request merged**